### PR TITLE
Load Lane and Store Lane instructions

### PR DESCRIPTION
**Pull request by Maratyszcza**
Introduction
=========

Both x86 SSE4.1 and ARM NEON instruction sets include instructions which load or store a **single** lane of a SIMD register, and this PR introduce equivalent instructions in WebAssembly SIMD. The single-lane load and store instructions cover several broad use-cases:
1. Non-contiguous loads and stores, when we need to combine elements from disjoint locations in a single SIMD vector, or scatter elements from a single SIMD vector into disjoint locations. 
2. Processing fewer than 128 bits of data. Sometimes the algorithm or data structures just don't expose enough data to utilize all 128 bits of a SIMD vector, but would nevertheless benefit from processing fewer elements in parallel (e.g. adding 8 bytes in one SIMD instruction rather than eight scalar instructions).

Load-Lane instructions complement the Load-Zero instructions (#237), but with different performance effects: a non-contiguous load sequence based on Load-Zero instructions would result in lower latency at the cost of throughput, while a sequence based on Load-Lane instructions would trade the low latency for higher throughput. Moreover, even when Load-Lane instructions are used, typically the first element is loaded with Load-Zero, and others are inserted with Load-Lane instructions.

Load-Lane instructions can be emulated via a combination of scalar loads and `replace_lane` instructions, and Store-Lane instructions can be emulated via a combination of `extract_lane` instructions and scalar stores. However, these emulation sequences are substantially less efficient that direct lane loads/stores with a SIMD register:
1. They need an extra register for the extracted scalar value.
2. They use an 2+ instructions where one would suffice on both ARM and x86.
3. For integer-type `replace_lane`/`extract_lane`, they involve moving values between SIMD and general-purpose registers, which comes at high latency and throughput cost.

Explicit instructions for loading and storing lanes alleviate all these concerns.

Applications
=========

- [BLIS (linear algebra)](https://github.com/flame/blis/blob/0645f239fbdf37ee9d2096ee3bb0e76b3302cfff/kernels/armv7a/3/bli_gemm_armv7a_int_d4x4.c#L247-L269)
- [XNNPACK (neural network inference)](https://github.com/google/XNNPACK/blob/9306ae024a4bba277a51ad0d31690e2183289d32/src/x32-pad/wasmsimd.c#L83)
- [Eigen (long vector SIMD library)](https://gitlab.com/libeigen/eigen/-/blob/14022f5eb5304e23ebe10284f07bb1387570dc5e/Eigen/src/Core/arch/NEON/PacketMath.h#L1914-1928)
- [TensorFlow (machine learning framework)](https://github.com/tensorflow/tensorflow/blob/e8598ce0454c440fca64e4ebc4aeedfa7afd5c97/tensorflow/core/kernels/quantized_resize_bilinear_op.cc#L156-L163)
- [gemmlowp (fixed-point matrix multiplication)](https://github.com/google/gemmlowp/blob/664f14a7754e6c78e7b063433c4d3e52fd843c09/internal/output_neon.h#L281-L288)

Mapping to Common Instruction Sets
===========================

This section illustrates how the new WebAssembly instructions can be lowered on common instruction sets. However, these patterns are provided only for convenience, compliant WebAssembly implementations do not have to follow the same code generation patterns.

x86/x86-64 processors with AVX instruction set
--------------------------------------------------

- **v128.load8_lane**
  - `y = v128.load8_lane(mem, x, lane)` is lowered to `VPINSRB xmm_y, xmm_x, [mem], lane`
- **v128.load16_lane**
  - `y = v128.load16_lane(mem, x, lane)` is lowered to `VPINSRW xmm_y, xmm_x, [mem], lane`
- **v128.load32_lane**
  - `y = v128.load32_lane(mem, x, lane)` is lowered to `VINSERTPS xmm_y, xmm_x, [mem], (lane << 4)`
- **v128.load64_lane**
  - `y = v128.load64_lane(mem, x, lane)` is lowered to:
    - `VMOVLPS xmm_y, xmm_x, [mem]` when `lane == 0`
    - `VMOVHPS xmm_y, xmm_x, [mem]` when `lane == 1`

- **v128.store8_lane**
  - `v128.store8_lane(mem, v, lane)` is lowered to `VPEXTRB [mem], xmm_v, lane`
- **v128.store16_lane**
  - `v128.store16_lane(mem, v, lane)` is lowered to `VPEXTRW [mem], xmm_v, lane`
- **v128.store32_lane**
  - `v128.store32_lane(mem, v, lane)` is lowered to:
    - `VMOVSS [mem], xmm_v` when `lane == 0`
    - `VEXTRACTPS [mem], xmm_v, lane` otherwise
- **v128.store64_lane**
  - `v128.store64_lane(mem, v, lane)` is lowered to:
    - `VMOVLPS [mem], xmm_v` when `lane == 0`
    - `VMOVHPS [mem], xmm_v` when `lane == 1`

x86/x86-64 processors with SSE4.1 instruction set
--------------------------------------------------

- **v128.load8_lane**
  - `y = v128.load8_lane(mem, x, lane)` is lowered to `MOVDQA xmm_y, xmm_x` + `PINSRB xmm_y, [mem], lane`
- **v128.load32_lane**
  - `y = v128.load32_lane(mem, x, lane)` is lowered to `MOVAPS xmm_y, xmm_x` + `INSERTPS xmm_y, [mem], (lane << 4)`

- **v128.store8_lane**
  - `v128.store8_lane(mem, v, lane)` is lowered to `PEXTRB [mem], xmm_v, lane`
- **v128.store16_lane**
  - `v128.store16_lane(mem, v, lane)` is lowered to `PEXTRW [mem], xmm_v, lane`
- **v128.store32_lane**
  - `v128.store32_lane(mem, v, lane)` is lowered to:
    - `MOVSS [mem], xmm_v` when `lane == 0`
    - `EXTRACTPS [mem], xmm_v, lane` otherwise

x86/x86-64 processors with SSE2 instruction set
--------------------------------------------------

- **v128.load8_lane**
  - `y = v128.load8_lane(mem, x, lane)` is lowered to:
    - `MOVD eax, xmm_x, (lane/2)` + `MOV al, byte [mem]` + `MOVDQA xmm_y, xmm_x` + `PINSRW xmm_y, eax, 0` when `lane == 0`
    - `MOVD eax, xmm_x, (lane/2)` + `MOV ah, byte [mem]` + `MOVDQA xmm_y, xmm_x` + `PINSRW xmm_y, eax, 0` when `lane == 1`
    - `PEXTRW eax, xmm_x, (lane/2)` + `MOV al, byte [mem]` + `MOVDQA xmm_y, xmm_x` + `PINSRW xmm_y, eax, (lane/2)`  when `lane is even` and `lane >= 2`
    - `PEXTRW eax, xmm_x, (lane/2)` + `MOV ah, byte [mem]` + `MOVDQA xmm_y, xmm_x` + `PINSRW xmm_y, eax, (lane/2)` when `lane is odd` and `lane >= 2`
- **v128.load16_lane**
  - `y = v128.load16_lane(mem, x, lane)` is lowered to `MOVDQA xmm_y, xmm_x` + `PINSRW xmm_y, [mem], lane`
- **v128.load32_lane**
  - `y = v128.load32_lane(mem, x, lane)` is lowered to:
    - `MOVAPS xmm_y, xmm_x` + `MOVSS xmm_tmp, [mem]` + `MOVSS xmm_y, xmm_tmp` when `lane == 0`
    - `MOVAPS xmm_y, xmm_x` + `PINSRW xmm_y, [mem], (lane*2)` + `PINSRW xmm_y, [mem+2], (lane*2+1)` otherwise
- **v128.load64_lane**
  - `y = v128.load64_lane(mem, x, lane)` is lowered to:
    - `MOVAPS xmm_y, xmm_x` + `MOVLPS xmm_y, [mem]` when `lane == 0`
    - `MOVAPS xmm_y, xmm_x` + `MOVHPS xmm_y, [mem]` when `lane == 1`

- **v128.store8_lane**
  - `v128.store8_lane(mem, v, lane)` is lowered to:
    - `MOVD eax, xmm_v` + `MOV byte [mem], al` when `lane == 0`
    - `MOVD eax, xmm_v` + `MOV byte [mem], ah` when `lane == 1`
    - `PEXTRW eax, xmm_v, (lane/2)` + `MOV byte [mem], al` when `lane is even` and `lane >= 2`
    - `PEXTRW eax, xmm_v, (lane/2)` + `MOV byte [mem], ah` when `lane is odd` and `lane >= 2`
- **v128.store16_lane**
  - `v128.store16_lane(mem, v, lane)` is lowered to `PEXTRW r_tmp, xmm_v, lane` + `MOV word [mem], r_tmp`
- **v128.store32_lane**
  - `v128.store32_lane(mem, v, lane)` is lowered to:
    - `MOVSS [mem], xmm_v` when `lane == 0`
    - `PSHUFD xmm_tmp, xmm_v, lane` + `MOVD [mem], xmm_tmp` otherwise
- **v128.store64_lane**
  - `v128.store64_lane(mem, v, lane)` is lowered to:
    - `MOVLPS [mem], xmm_v` when `lane == 0`
    - `MOVHPS [mem], xmm_v` when `lane == 1`

ARM64 processors
--------------------------------------------------

- **v128.load8_lane**
  - `y = v128.load8_lane(mem, x, lane)` is lowered to `MOV Qy, Qx` + `LD1 {Vy.B}[lane], [Xmem]`
- **v128.load16_lane**
  - `y = v128.load16_lane(mem, x, lane)` is lowered to `MOV Qy, Qx` + `LD1 {Vy.H}[lane], [Xmem]`
- **v128.load32_lane**
  - `y = v128.load32_lane(mem, x, lane)` is lowered to `MOV Qy, Qx` + `LD1 {Vy.S}[lane], [Xmem]`
- **v128.load64_lane**
  - `y = v128.load64_lane(mem, x, lane)` is lowered to `MOV Qy, Qx` + `LD1 {Vy.D}[lane], [Xmem]`

- **v128.store8_lane**
  - `v128.store8_lane(mem, v, lane)` is lowered to `ST1 {Vv.B}[lane], [Xmem]`
- **v128.store16_lane**
  - `v128.store16_lane(mem, v, lane)` is lowered to `ST1 {Vv.H}[lane], [Xmem]`
- **v128.store32_lane**
  - `v128.store32_lane(mem, v, lane)` is lowered to `ST1 {Vv.S}[lane], [Xmem]`
- **v128.store64_lane**
  - `v128.store64_lane(mem, v, lane)` is lowered to `ST1 {Vv.D}[lane], [Xmem]`

ARMv7 processors with NEON instruction set
--------------------------------------------------

- **v128.load8_lane**
  - `y = v128.load8_lane(mem, x, lane)` is lowered to:
    - `VMOV Qy, Qx` + `VLD1.8 {Dy_lo[lane]}, [Xmem]` when `lane < 8`
    - `VMOV Qy, Qx` + `VLD1.8 {Dy_hi[(lane-8)]}, [Xmem]` when `lane >= 8`
- **v128.load16_lane**
  - `y = v128.load16_lane(mem, x, lane)` is lowered to:
    - `VMOV Qy, Qx` + `VLD1.16 {Dy_lo[lane]}, [Xmem]` when `lane < 4`
    - `VMOV Qy, Qx` + `VLD1.16 {Dy_hi[(lane-4)]}, [Xmem]` when `lane >= 4`
- **v128.load32_lane**
  - `y = v128.load32_lane(mem, x, lane)` is lowered to:
    - `VMOV Qy, Qx` + `VLD1.32 {Dy_lo[lane]}, [Xmem]` when `lane < 2`
    - `VMOV Qy, Qx` + `VLD1.32 {Dy_hi[(lane-2)]}, [Xmem]` when `lane >= 2`
- **v128.load64_lane**
    - `VMOV Dy_hi, Dx_hi` + `VLD1.64 {Dy_lo}, [Xmem]` when `lane == 0`
    - `VMOV Dy_lo, Dx_lo` + `VLD1.64 {Dy_hi}, [Xmem]` when `lane == 1`

- **v128.store8_lane**
  - `v128.store8_lane(mem, v, lane)` is lowered to `ST1 {Vv.B}[lane], [Xmem]`
  - `v128.store8_lane(mem, v, lane)` is lowered to:
    - `VST1.8 {Dy_lo[lane]}, [Xmem]` when `lane < 8`
    - `VST1.8 {Dy_hi[(lane-8)]}, [Xmem]` when `lane >= 8`
- **v128.store16_lane**
  - `v128.store16_lane(mem, v, lane)` is lowered to:
    - `VST1.16 {Dy_lo[lane]}, [Xmem]` when `lane < 4`
    - `VST1.16 {Dy_hi[(lane-4)]}, [Xmem]` when `lane >= 4`
- **v128.store32_lane**
    - `VST1.32 {Dy_lo[lane]}, [Xmem]` when `lane < 2`
    - `VST1.32 {Dy_hi[(lane-2)]}, [Xmem]` when `lane >= 2`
- **v128.store64_lane**
  - `v128.store64_lane(mem, v, lane)` is lowered to:
    - `VST1.64 {Dy_lo}, [Xmem]` when `lane == 0`
    - `VST1.64 {Dy_hi}, [Xmem]` when `lane == 1`

### CODE REVIEW

**Comment by Maratyszcza** [comment]
@tlively Please advice which opcodes could be used for these instructions

**Comment by tlively** [comment]
We're basically out of opcode space, so you should just append these to the end for now. When we did the last renumbering, we thought we were essentially done adding new opcodes, so if we end up including all these newly proposed instructions, we'll probably have to do yet another renumbering.

**Comment by Maratyszcza** [comment]
@tlively There isn't enough space in the end, inserted at `0x58`-`0x5f` instead.

**Comment by tlively** [comment]
Opcodes are encoded as ULEB128s, so it's totally fine to use numbers above 0xff.

**Comment by Maratyszcza** [comment]
@tlively Good point! Do you prefer to move them to the end, or is it fine to leave the opcodes as is.

**Comment by tlively** [comment]
No, that's ok. Their current position looks fine. Thanks!

File: proposals/simd/BinarySIMD.md [hunk 1]
```diff
@@ -30,195 +30,203 @@ In the description below, `ImmLaneIdx{I}` indicates the maximum value of the byt
...
+| `f64x2.gt`                 |    `0x4a`| -                        |
+| `f64x2.le`                 |    `0x4b`| -                        |
+| `f64x2.ge`                 |    `0x4c`| -                        |
+| `v128.not`                 |    `0x4d`| -                        |
+| `v128.and`                 |    `0x4e`| -                        |
+| `v128.andnot`              |    `0x4f`| -                        |
+| `v128.or`                  |    `0x50`| -                        |
+| `v128.xor`                 |    `0x51`| -                        |
+| `v128.bitselect`           |    `0x52`| -                        |
+| `v128.load8_splat`         |    `0x58`| m:memarg, i:ImmLaneIdx16 |
```
Comment (ngzhian) [inline proposals/simd/BinarySIMD.md:118]: typo: s/splat/lane
same for the 3 other loads right below.

Comment (Maratyszcza) [reply proposals/simd/BinarySIMD.md]: Done

**Comment by ngzhian** [comment]
Prototyped for x64 in https://crrev.com/c/2444578, should see it in canary by tomorrow. @Maratyszcza

**Comment by tlively** [comment]
These instructions have now landed in both LLVM and Binaryen so they will be ready to use in tip-of-tree Emscripten (usable via `emsdk install tot && emsdk activate tot`) in a few hours. The builtin functions for these instructions are `__builtin_wasm_{load,store}{8,16,32,64}_lane(lane_t*, vec_t, lane_index)`. Since these are still prototypes, there are no corresponding wasm_simd128.h instrinsics yet.

**Comment by Maratyszcza** [comment]
I evaluated the performance impact of these instructions by modifying WebAssembly SIMD microkernels for Sigmoid operator in [XNNPACK library](https://github.com/google/XNNPACK) of neural network operators to use `v128.load32_lane` instruction, and the results are below:

Processor | Performance with WAsm SIMD + `v128.load32_zero` + `v128.load32_lane` (this PR) | Performance with WAsm SIMD + `v128.load32_zero` (PR #237) | Speedup |
-- | -- | -- | --
Intel Xeon W-2135 | 6.57 GB/s | 6.35 GB/s | 3% |
AMD PRO A10-8700B | 3.19 GB/s | 3.10 GB/s | 3% |
Snapdragon 670 (Pixel 3a) | 1.55 GB/s | 1.49 GB/s | 4% |

The code modifications can be seen in google/XNNPACK#1016 (for the baseline version with `v128.load32_zero`) and in google/XNNPACK#1199 (for the optimized version with both `v128.load32_zero` and `v128.load32_lane`).

**Comment by Maratyszcza** [comment]
Attn @abrown

**Comment by omnisip** [comment]
Discussed in (#402 12/22/2020 Sync Meeting).

Provisional Voting Results: 
2 SF - 5 F - 1 N

Minutes are [here](https://docs.google.com/document/d/1Tnf-fvRcCVj_vv8CjAN_qUvj4Wa57JnPa8ghGi4_C-s/edit#).

File: proposals/simd/BinarySIMD.md [hunk 2]
```diff
@@ -30,210 +30,218 @@ In the description below, `ImmLaneIdx{I}` indicates the maximum value of the byt
...
+| `v128.xor`                  |    `0x51`| -                        |
+| `v128.bitselect`            |    `0x52`| -                        |
+| `v128.load8_lane`           |    `0x58`| m:memarg, i:ImmLaneIdx16 |
+| `v128.load16_lane`          |    `0x59`| m:memarg, i:ImmLaneIdx16 |
+| `v128.load32_lane`          |    `0x5a`| m:memarg, i:ImmLaneIdx16 |
+| `v128.load64_lane`          |    `0x5b`| m:memarg, i:ImmLaneIdx16 |
+| `v128.store8_lane`          |    `0x5c`| m:memarg, i:ImmLaneIdx16 |
+| `v128.store16_lane`         |    `0x5d`| m:memarg, i:ImmLaneIdx16 |
+| `v128.store32_lane`         |    `0x5e`| m:memarg, i:ImmLaneIdx16 |
+| `v128.store64_lane`         |    `0x5f`| m:memarg, i:ImmLaneIdx16 |
```
Comment (ngzhian) [inline proposals/simd/BinarySIMD.md:125]: `ImmLaneIdx16` needs to be updated for {load,store}_{16,32,64}

Comment (Maratyszcza) [reply proposals/simd/BinarySIMD.md]: Fixed

**Review by tlively — APPROVED** [review]

**Review by dtig — CHANGES_REQUESTED** [review]
Please add the new operations to ImplementationStatus.md as well.

**Comment by Maratyszcza** [comment]
@dtig Done

**Review by dtig — APPROVED** [review]

**Comment by Maratyszcza** [comment]
Rebased on top of merged PRs

**Comment by Maratyszcza** [comment]
Rebased once again
